vdev/src/macros.rs

```diff
@@ -0,0 +1,50 @@
+macro_rules! display {
+    ($($arg:tt)*) => {{
+        use owo_colors::OwoColorize;
+        println!(
+            "{}",
+            <std::string::String as AsRef<str>>::as_ref(&format!($($arg)*))
+                .if_supports_color(owo_colors::Stream::Stdout, |text| text.bold())
+        );
+    }};
+}
+
+macro_rules! display_critical {
+    ($($arg:tt)*) => {{
+        use owo_colors::OwoColorize;
+        eprintln!(
+            "{}",
+            <std::string::String as AsRef<str>>::as_ref(&format!($($arg)*))
+                .if_supports_color(owo_colors::Stream::Stderr, |text| text.bright_red())
+        );
+    }};
+}
+
+macro_rules! define_display_macro {
+    // https://github.com/rust-lang/rust/issues/35853#issuecomment-415993963
+    // https://github.com/rust-lang/rust/issues/83527#issuecomment-1281176235
+    ($name:ident, $level:ident, $style:ident, $d:tt) => (
+        #[allow(unused_macros)]
+        macro_rules! $name {
+            ($d($d arg:tt)*) => {
+                if log::Level::$level <= *$crate::app::verbosity() {
+                    use owo_colors::OwoColorize;
+                    eprintln!(
+                        "{}",
+                        <std::string::String as AsRef<str>>::as_ref(&format!($d($d arg)*))
+                            .if_supports_color(owo_colors::Stream::Stderr, |text| text.$style())
+                    );
+                }
+            };
+        }
+    );
+}
+
+// Simply bold rather than bright white for terminals with white backgrounds
+define_display_macro!(display_trace, Trace, bold, $);
+define_display_macro!(display_debug, Debug, bold, $);
+define_display_macro!(display_info, Info, bold, $);
+define_display_macro!(display_success, Info, bright_cyan, $);
+define_display_macro!(display_waiting, Info, bright_magenta, $);
+define_display_macro!(display_warning, Warn, bright_yellow, $);
+define_display_macro!(display_error, Error, bright_red, $);
```

Comment:
I'm not a big fan of this macro-generating-macro. Could this be done more simply with a semi-magic parameter on a single macro to enable usage like:
```rust
    display!("Plain text");
    display!(critical, "Run!");
    display!(trace, "This line happened.");
```
This can be achieved by using the literal words `critical` etc in the macro rules pattern matcher.

Reply:
I could if you really want but that would be very unergonomic and not as nice of a dev UX. Others do this too:

- https://github.com/rust-lang/log/blob/master/src/macros.rs
- https://github.com/SergioBenitez/Rocket/blob/master/core/lib/src/log.rs